drivers/disk/ftl_dhara.c

```diff
@@ -0,0 +1,424 @@
+
+		ret = dhara_map_write(&ctx->dhara_map, sector, buffer, &err);
+		if (ret != 0) {
+			LOG_ERR("dhara_map_write failed with error %d", err);
+			k_sem_give(&ctx->lock);
+			return -EIO;
+		}
+	}
+
+	k_sem_give(&ctx->lock);
+
+	return 0;
+}
+
+static int disk_ftl_access_ioctl(struct disk_info *disk, uint8_t cmd, void *buff)
+{
+	struct disk_ftl_data *ctx = CONTAINER_OF(disk, struct disk_ftl_data, info);
+	dhara_error_t err;
+	int ret;
+
+	switch (cmd) {
+	case DISK_IOCTL_GET_SECTOR_COUNT:
+		k_sem_take(&ctx->lock, K_FOREVER);
+		*(uint32_t *)buff = dhara_map_capacity(&ctx->dhara_map);
+		k_sem_give(&ctx->lock);
+		break;
+
+	case DISK_IOCTL_GET_SECTOR_SIZE:
+		*(uint32_t *)buff = ctx->page_size;
+		break;
+
+	case DISK_IOCTL_GET_ERASE_BLOCK_SZ:
+		*(uint32_t *)buff = ctx->block_size;
+		break;
+
+	case DISK_IOCTL_CTRL_SYNC:
+	case DISK_IOCTL_CTRL_DEINIT:
+		k_sem_take(&ctx->lock, K_FOREVER);
+		ret = dhara_map_sync(&ctx->dhara_map, &err);
+		k_sem_give(&ctx->lock);
+		if (ret != 0) {
+			LOG_ERR("dhara_map_sync failed with error %d", err);
+			return -EIO;
+		}
+		break;
+
+	case DISK_IOCTL_CTRL_INIT:
+		return disk_ftl_access_init(disk);
+
+	default:
+		LOG_ERR("Unsupported ioctl command %u", cmd);
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+static int disk_ftl_init(const struct device *dev)
+{
+	struct disk_ftl_data *dev_data = dev->data;
+
+	return disk_access_register(&dev_data->info);
+}
+
+static const struct disk_operations disk_ftl_ops = {
+	.init = disk_ftl_access_init,
+	.status = disk_ftl_access_status,
+	.read = disk_ftl_access_read,
+	.write = disk_ftl_access_write,
+	.ioctl = disk_ftl_access_ioctl,
+};
+
+#define PARTITION_PHANDLE(n) DT_PHANDLE(DT_DRV_INST(n), partition)
+
+#define DISK_FTL_INIT(n)                                                                           \
+	static uint8_t disk_ftl_page_buffer_##n[DT_INST_PROP(n, buffer_size)];                     \
+                                                                                                   \
+	static uint8_t disk_ftl_dhara_buffer_##n[DT_INST_PROP(n, buffer_size)];                    \
+                                                                                                   \
+	static struct disk_ftl_data disk_ftl_data_##n = {                                          \
+		.info =                                                                            \
+			{                                                                          \
+				.name = DT_INST_PROP(n, disk_name),                                \
+				.ops = &disk_ftl_ops,                                              \
+			},                                                                         \
+		.partition_id = DT_FIXED_PARTITION_ID(PARTITION_PHANDLE(n)),                       \
+		.partition_offset = DT_REG_ADDR(PARTITION_PHANDLE(n)),                             \
+		.partition_size = DT_REG_SIZE(PARTITION_PHANDLE(n)),                               \
```

Comment:
Use FIXED_PARTITION macros here, and instead of ID, use the macro, FIXED_PARTITION, that directly takes pointer to `struct flash_area` object instance serving the selected partition.